cloudinit/osys/windows/network.py

```diff
@@ -0,0 +1,415 @@
+                    curr_addr.Next, ctypes.POINTER(
+                        iphlpapi.IP_ADAPTER_ADDRESSES))
+
+            finally:
+                kernel32.HeapFree(proc_heap, 0, p)
+                ws2_32.WSACleanup()
+
+        return interfaces
+
+    def set_static_network_config(self, mac_address, address, netmask,
+                                  gateway, dnsnameservers):
+        conn = wmi.WMI(moniker='//./root/cimv2')
+
+        q = conn.query("SELECT * FROM Win32_NetworkAdapter WHERE "
+                       "MACAddress = '{}'".format(mac_address))
+        if not len(q):
+            raise Exception("Network adapter not found")
+
+        adapter_config = q[0].associators(
+            wmi_result_class='Win32_NetworkAdapterConfiguration')[0]
+
+        LOG.debug("Setting static IP address")
+        (ret_val, ) = adapter_config.EnableStatic([address], [netmask])
+        if ret_val > 1:
+            raise Exception(
+                "Cannot set static IP address on network adapter")
+        reboot_required = (ret_val == 1)
+
+        if gateway:
+            LOG.debug("Setting static gateways")
+            (ret_val, ) = adapter_config.SetGateways([gateway], [1])
+            if ret_val > 1:
+                raise Exception("Cannot set gateway on network adapter")
+            reboot_required = reboot_required or ret_val == 1
+
+        if dnsnameservers:
+            LOG.debug("Setting static DNS servers")
+            (ret_val,) = adapter_config.SetDNSServerSearchOrder(dnsnameservers)
+            if ret_val > 1:
+                raise Exception("Cannot set DNS on network adapter")
+            reboot_required = reboot_required or ret_val == 1
+
+        return reboot_required
+
+    def set_hostname(self, hostname):
+        ret_val = kernel32.SetComputerNameExW(
+            _ComputerNamePhysicalDnsHostname,
+            six.text_type(hostname))
+        if not ret_val:
+            raise Exception("Cannot set host name")
+
+    @staticmethod
+    def _get_fw_protocol(protocol):
+        if protocol == _PROTOCOL_TCP:
+            fw_protocol = _FW_IP_PROTOCOL_TCP
+        elif protocol == _PROTOCOL_UDP:
+            fw_protocol = _FW_IP_PROTOCOL_UDP
+        else:
+            raise NotImplementedError("Unsupported protocol")
+        return fw_protocol
+
+    def firewall_create_rule(self, name, port, protocol, allow=True):
+        if not allow:
+            raise NotImplementedError()
+
+        fw_port = client.Dispatch("HNetCfg.FWOpenPort")
+        fw_port.Name = name
+        fw_port.Protocol = self._get_fw_protocol(protocol)
+        fw_port.Port = port
+        fw_port.Scope = _FW_SCOPE_ALL
+        fw_port.Enabled = True
+
+        fw_mgr = client.Dispatch("HNetCfg.FwMgr")
+        fw_profile = fw_mgr.LocalPolicy.CurrentProfile
+        fw_profile = fw_profile.GloballyOpenPorts.Add(fw_port)
+
+    def firewall_remove_rule(self, _, port, protocol, allow=True):
+        if not allow:
```

Comment:
Hmmm, seems like something other than ``NotImplementedError`` should be raised here since it actually is implemented but not allowed...